crates/ruma-events/src/sticky.rs

```diff
@@ -0,0 +1,86 @@
+//! MSC4354 Sticky events types and utils.
+
+use js_int::UInt;
+use serde::{Deserialize, Serialize};
+
+/// Sticky duration in milliseconds.
+/// Valid values are the integer range 0-3600000 (1 hour)
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Deserialize, Serialize)]
+pub struct StickyDurationMs(u32);
+
+impl StickyDurationMs {
+    const MAX: u32 = 3_600_000;
+
+    /// Creates a new `StickyDurationMs` if `v` is within `[0, 1h]`.
+    ///
+    /// Returns an error if `v` exceeds the maximum.
+    pub fn new(v: u32) -> Result<Self, &'static str> {
+        if v <= Self::MAX {
+            Ok(Self(v))
+        } else {
+            Err("out of range [0, 3_600_000]")
+        }
+    }
+
+    /// Creates a `DurationMs` by clamping `v` into `[0, 1h]`.
+    pub fn new_wrapping<T: Into<u64>>(v: T) -> Self {
+        let v = v.into();
+        let clamped = v.min(Self::MAX as u64) as u32;
+        Self(clamped)
+    }
+
+    /// Get the value as a `u32`.
+    pub fn get(self) -> u32 {
+        self.into()
+    }
+}
+
+impl From<StickyDurationMs> for u32 {
+    fn from(d: StickyDurationMs) -> Self {
+        d.0
+    }
+}
+
+/// Message events can be annotated with a new top-level sticky object,
+/// which MUST have a duration_ms, which is the number of milliseconds for the event to be
+/// sticky.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct StickyObject {
+    // private, use `capped_duration_ms()`
+    duration_ms: UInt,
+}
+
+impl StickyObject {
+    /// Valid values are the integer range 0-3600000 (1 hour)
+    pub fn clamped_duration_ms(&self) -> StickyDurationMs {
```

Comment:
This API looks like we can use durations that are out of bounds. The MSC says the opposite (emphasis mine):

> The presence of `sticky.duration_ms` **with a valid value** makes the event “sticky”.

So an invalid value should be ignored.

I believe that the correct way to deal with this is:

1. Implement custom deserialization for `StickyDurationMs` that returns an error when the value is out of bounds.
2. Use `StickyDurationMs` for this field and make it public.
3. When using `StickyObject`, we should use the equivalent of `ruma_common::serde::default_on_error`, to ignore the object when it fails to deserialize. However this is not possible currently with the macro.